### PR TITLE
add tracing interceptor to persist trace id to caller

### PR DESCRIPTION
**Pull request by javidang**

### CODE REVIEW

File: core/src/main/java/io/token/TokenClient.java [hunk 1]
```diff
@@ -646,6 +647,29 @@ public MemberRecoveryOperation getRecoveryAuthorizationBlocking(
         return getCountries(provider).blockingSingle();
     }
 
+    /**
+     * Wraps provided response message with associated trace id.
```
Comment (ab666) [inline core/src/main/java/io/token/TokenClient.java:651]: new line before params

File: core/src/main/java/io/token/TokenClient.java [hunk 2]
```diff
@@ -646,6 +647,29 @@ public MemberRecoveryOperation getRecoveryAuthorizationBlocking(
...
+     */
+    public static <T> Observable<WithTraceId<T>> wrapWithTraceId(Observable<T> observable) {
+        return observable.map(message -> {
+            String traceId = getTraceId();
+            return new WithTraceId<>(message, traceId);
+        });
+    }
+
+    /**
+     * Fetches trace id from ThreadLocal and removes it.
```
Comment (ab666) [inline core/src/main/java/io/token/TokenClient.java:664]: same here

File: core/src/main/java/io/token/WithTraceId.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package io.token;
+
+import jdk.internal.jline.internal.Nullable;
+
+/**
+ * Wrapper class around {@link com.google.protobuf.Message} that includes trace id.
```
Comment (ab666) [inline core/src/main/java/io/token/WithTraceId.java:28]: and here

File: core/src/main/java/io/token/rpc/TracingInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.token.rpc.interceptor.SimpleInterceptor;
+import io.token.rpc.util.Tracing;
+
+import javax.annotation.Nullable;
+
+/**
+ * Interceptor to fetch token-trace-id from grpc call metadata.
```
Comment (ab666) [inline core/src/main/java/io/token/rpc/TracingInterceptor.java:35]: well, it looks like everywhere...

**Comment by javidang** [comment]
> Not sure if I like the `WithTraceId`
> How about TracableResult?

@ab666 , sure. What about function name then? `traceResult` ?

**Comment by ab666** [comment]
trace(...) or traceable(...)

**Pull request merged**